Adding Input component and updating TextFieldBase (#17664)
MetaMask/metamask-extension: ui/components/component-library/input/README.mdx

import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import { Input } from './input';

# Input

`Input` is a light-weight borderless input component used inside of [TextField](/docs/components-componentlibrary-textfield--default-story#textfield)

<Canvas>
  <Story id="components-componentlibrary-input--default-story" />
</Canvas>

## Props

The `Input` accepts all props below as well as all [Box](/docs/components-ui-box--default-story#props) component props

<ArgsTable of={Input} />

### Type

Use the `type` prop to change the type of input.

Possible types include:

- `INPUT_TYPES.TEXT`
- `INPUT_TYPES.NUMBER`
- `INPUT_TYPES.PASSWORD`
- `INPUT_TYPES.SEARCH`

> Defaults to `text`.

Author: Update to 
```suggestion
Defaults to `INPUT_TYPES.TEX`.
```

<Input type={INPUT_TYPES.PASSWORD} />
<Input type={INPUT_TYPES.SEARCH} />
```

### Ref

Use the `ref` prop to access the ref of the `<input />` html element of `Input`. This is useful for focusing the input from a button or other component.

<Canvas>
  <Story id="components-componentlibrary-input--ref" />
</Canvas>

```jsx
import { Button, Input } from '../../component-library';

const inputRef = useRef(null);
const [value, setValue] = useState('');
const handleOnClick = () => {
  inputRef.current.focus();
};
const handleOnChange = (e) => {
  setValue(e.target.value);
};

<Input
  ref={inputRef}
  value={value}
  onChange={handleOnChange}
/>
<Button marginLeft={1} onClick={handleOnClick}>
  Edit
</Button>
```

### Auto Complete

Use the `autoComplete` prop to set the autocomplete html attribute. It allows the browser to predict the value based on earlier typed values.

<Canvas>
  <Story id="components-componentlibrary-input--auto-complete" />
</Canvas>

```jsx
import { Input } from '../../component-library';

<Input type={INPUT_TYPES.PASSWORD} autoComplete />;
```

### Auto Focus

Use the `autoFocus` prop to focus the `Input` during the first mount

<Canvas>

>   <Story id="components-componentlibrary-input--auto-focus" />

Reviewer: .github/workflows/blank.yml

>   <Story id="components-componentlibrary-input--auto-focus" />

Author: Yeah good point. It's very annoying. Removed here: https://github.com/MetaMask/metamask-extension/pull/17664/commits/6a1e1b6fe5938d058f48ae933139f3874a79680f

https://user-images.githubusercontent.com/8112138/218845047-9274ef28-4175-48ce-9e84-1f5a70e6c14f.mov

</Canvas>

```jsx
import { Input } from '../../component-library';

<Input autoFocus />;
```

### Default Value

> Use the `defaultValue` prop to set the default value of the `Input`

Author: Update to 
```suggestion
Use the `defaultValue` prop to set the default value of the `Input`. Use for uncontrolled inputs
```

<Canvas>
  <Story id="components-componentlibrary-input--default-value" />
</Canvas>

```jsx
import { Input } from '../../component-library';

<Input defaultValue="default value" />;
```

### Disabled

Use the `disabled` prop to set the disabled state of the `Input`

<Canvas>
  <Story id="components-componentlibrary-input--disabled" />
</Canvas>

```jsx
import { Input } from '../../component-library';

<Input disabled />;
```

### Error

> Use the `error` prop to set `aria-invalid="true"`. This helps with screen readers for accessibility.

Author: Update to
```suggestion
Use the `error` prop to set `aria-invalid="true"`. This helps with screen readers for accessibility. There is no visual indicator for `error` this should be handled in the parent component.
```

<Canvas>
  <Story id="components-componentlibrary-input--error-story" />
</Canvas>

```jsx
import { Input } from '../../component-library';

<Input error />;
```

### Max Length

Use the `maxLength` prop to set the maximum allowed input characters for the `Input`

<Canvas>
  <Story id="components-componentlibrary-input--max-length" />
</Canvas>

```jsx
import { Input } from '../../component-library';

<Input maxLength={10} />;
```

### Read Only

> Use the `readOnly` prop to set the `Input` to read only. When `readOnly` is true `Input` will not have a focus state.

Author: Update to
```suggestion
Use the `readOnly` prop to set the `Input` to read only
```

<Canvas>
  <Story id="components-componentlibrary-input--read-only" />
</Canvas>

```jsx
import { Input } from '../../component-library';

<Input readOnly />;
```

### Required

> Use the `required` prop to set the html `required` attribute. There is no visual difference to the `Input` when required is true but is used by the browser api.

Author: Update to
```suggestion
Use the `required` prop to set the html `required` attribute used by the browser api. There is no visual indicator for `required` this should be handled in the parent component.  
```

<Canvas>
  <Story id="components-componentlibrary-input--required" />
</Canvas>

```jsx
import { Input } from '../../component-library';

> // No visual difference. Used by the browser api

Author: ```suggestion
// No visual indicator. Used by the browser api
```
